src/ncp/ncp_uart.cpp

```diff
@@ -47,138 +47,163 @@ extern "C" void otNcpInit(void)
+    sNcpUart->TxFrameBufferHasData();
+}
 
-    return errorCode;
+void NcpUart::TxFrameBufferHasData(void)
+{
+    if (mUartBuffer.IsEmpty())
+    {
+        EncodeAndSendToUart();
+    }
 }
 
-ThreadError
-NcpUart::OutboundFrameFeedMessage(Message &message)
+// This method encodes a frame from the tx frame buffer (mTxFrameBuffer) into the uart buffer and sends it over uart.
+// If the uart buffer gets full, it sends the current encoded portion. This method remembers current state, so on
+// sub-sequent calls, it restarts encoding the bytes from where it left of in the frame .
+void NcpUart::EncodeAndSendToUart(void)
 {
-    ThreadError errorCode;
-    uint16_t inLength;
-    uint8_t inBuf[16];
+    uint16_t len;
 
-    for (int offset = 0; offset < message.GetLength(); offset += sizeof(inBuf))
+    while (!mTxFrameBuffer.IsEmpty())
     {
-        (void) OutboundFrameGetRemaining();
-        inLength = message.Read(offset, sizeof(inBuf), inBuf);
+        switch (mState)
+        {
+        case kStartingFrame:
 
-        errorCode = OutboundFrameFeedData(inBuf, inLength);
+            SuccessOrExit(mFrameEncoder.Init(mUartBuffer));
 
-        if (errorCode != kThreadError_None)
-        {
-            break;
-        }
-    }
+            mTxFrameBuffer.OutFrameBegin();
 
-    return errorCode;
-}
+            mState = kEncodingFrame;
 
-ThreadError
-NcpUart::OutboundFrameSend(void)
-{
-    ThreadError errorCode;
+            while (!mTxFrameBuffer.OutFrameHasEnded())
+            {
+                mByte = mTxFrameBuffer.OutFrameReadByte();
 
-    errorCode = mFrameEncoder.Finalize(mSendFrame);
+            case kEncodingFrame:
 
-    if (errorCode == kThreadError_None)
-    {
-        // We go ahead and set this to `true` here in case
-        // `otPlatUartSend()` ends up directly calling
-        // `otPlatUartSendDone()`.
-        mSending = true;
+                SuccessOrExit(mFrameEncoder.Encode(mByte, mUartBuffer));
+            }
+
+            mTxFrameBuffer.OutFrameRemove();
+
+            // Notify the super/base class that there is space available in tx frame buffer for a new frame.
+            super_t::HandleSpaceAvailableInTxBuffer();
+
+            mState = kFinalizingFrame;
 
-        errorCode = otPlatUartSend(mSendFrame.GetBuffer(), mSendFrame.GetLength());
+        case kFinalizingFrame:
+
+            SuccessOrExit(mFrameEncoder.Finalize(mUartBuffer));
+
+            mState = kStartingFrame;
+        }
     }
 
-    if (errorCode != kThreadError_None)
+exit:
+    len = mUartBuffer.GetLength();
+
+    if (len > 0)
     {
-        mSending = false;
+        otPlatUartSend(mUartBuffer.GetBuffer(), len);
     }
-
-    return errorCode;
 }
 
 extern "C" void otPlatUartSendDone(void)
 {
-    sNcpUart->SendDoneTask();
+    sNcpUart->PostUartSendDoneTasklet();
 }
 
-void NcpUart::SendDoneTask(void)
+void NcpUart::PostUartSendDoneTasklet(void)
 {
-    mSending = false;
+    mUartSendDoneTask.Post();
```

Comment:
Do we still need to use a `Tasklet` here?  `otPlatUartSendDone` should already be called in OpenThread context.